katdal/applycal.py

```diff
@@ -51,25 +93,48 @@ def get_cal_product(cache, attrs, product):
     return CategoricalData(values, events)
 
 
-def calc_delay_correction(sensor, index, freqs):
+def calc_delay_correction(sensor, index, data_freqs):
     """Calculate correction sensor from delay calibration solution sensor.
 
     Given the delay calibration solution `sensor`, this extracts the delay time
     series of the input specified by `index` (in the form (pol, ant)) and
     builds a categorical sensor for the corresponding complex correction terms
-    (channelised by `freqs`).
+    (channelised by `data_freqs`).
 
     Invalid delays (NaNs) are replaced by zeros, since bandpass calibration
     still has a shot at fixing any residual delay.
     """
     delays = [np.nan_to_num(sensor[n][index]) for n in sensor.events[:-1]]
     # Delays returned by cal pipeline are already corrections (no minus needed)
-    corrections = [np.exp(2j * np.pi * d * freqs).astype('complex64')
+    corrections = [np.exp(2j * np.pi * d * data_freqs).astype('complex64')
                    for d in delays]
+    corrections = [ComparableArrayWrapper(c) for c in corrections]
+    return CategoricalData(corrections, sensor.events)
+
+
+def calc_bandpass_correction(sensor, index, data_freqs, cal_freqs):
+    """Calculate correction sensor from bandpass calibration solution sensor.
+
+    Given the bandpass calibration solution `sensor`, this extracts the time
+    series of bandpasses (channelised by `cal_freqs`) for the input specified
+    by `index` (in the form (pol, ant)) and builds a categorical sensor for
+    the corresponding complex correction terms (channelised by `data_freqs`).
+
+    Invalid solutions (NaNs) are replaced by linear interpolations over
+    frequency (separately for magnitude and phase), as long as some channels
+    have valid solutions.
+    """
+    corrections = []
+    for n in sensor.events[:-1]:
+        bp = sensor[n][(slice(None),) + index]
+        valid = np.isfinite(bp)
+        if valid.any():
+            bp = complex_interp(data_freqs, cal_freqs[valid], bp[valid])
```

Comment:
Done.

FWIW, extending the edge value is not all bad, since the gain will be inverted to become a correction. By overestimating the gain we are preventing a corresponding blowup in the correction, which could have boosted these noisy flanks to unacceptable levels.

This makes me think we need a cap on the gain correction magnitude for practical reasons...